src/libertem/web/server.py

```diff
@@ -130,6 +133,43 @@ def main(host, port, numeric_level, event_registry, shared_state):
     return app
 
 
+def _confirm_exit(shared_state, loop):
+    log.info('interrupted, ')
+    sys.stdout.write("Shutdown libertem server (y/[n])? ")
+    sys.stdout.flush()
+    r, w, x = select.select([sys.stdin], [], [], 5)
+    if r:
+        line = sys.stdin.readline()
+        if line.lower().startswith('y'):
+            log.critical("Shutdown confirmed")
+            # schedule stop on main thread
+            loop.add_callback_from_signal(
+                lambda: asyncio.ensure_future(do_stop(shared_state))
+            )
+            return
+    else:
+        print('No answer for 5s: ')
+    print('Resuming operation ...')
+    # set it back to original SIGINT handler
+    loop.add_callback_from_signal(partial(handle_signal, shared_state))
+
+
+def _handle_exit(signum, frame, shared_state):
+    loop = tornado.ioloop.IOLoop.current()
+    # register more forceful signal handler for ^C^C case
+    signal.signal(signal.SIGINT, partial(sig_exit, shared_state=shared_state))
+    thread = threading.Thread(target=partial(_confirm_exit, shared_state, loop))
+    thread.daemon = True
+    thread.start()
+
+
+def handle_signal(shared_state):
+    if not sys.platform.startswith('win'):
```

Comment:
[jupyter](https://github.com/jupyter/notebook/blob/e577b6c40c194d70c058f30aee33ba4fca5502b9/notebook/notebookapp.py#L1578) also check for `sys.stdin` and `sys.stdin.isatty()` , do we need this ?

Reply:
Yes: if the server is running in the background, detached from `stdin`, we can't ask the user for confirmation.

I'll do a more thorough review beginning next week